### PR TITLE
Add perplexity metric

### PR DESCRIPTION
**Pull request by JayOrten**
- Remove the parameter 'value_embed_dim', and related assertion; it isn't being used anywhere in the code, seems vestigial from past refactoring.
- Add perplexity logging in validation step of model training.
- Add a parameter to the config file for precision, as well as passes that parameter to lightning during training, to use lower bit precision.

### CODE REVIEW

File: src/train_model.py [hunk 1]
```diff
@@ -19,10 +19,6 @@
 from torchinfo import summary as model_summary
 from utils import Struct
 
-# Allow torch to run float32 matrix multiplications in lower precision for
-# better performance while training if hardware is capable
-torch.backends.cuda.matmul.allow_tf32 = True
```
Comment (KimballNJardine) [inline src/train_model.py:24]: I assume that precision in trainer covers this instead, right?

File: src/models.py [hunk 2]
```diff
@@ -236,6 +248,8 @@ def validation_step(self, batch: Tensor, batch_idx: int):
         # Calculate loss
```
Comment (KimballNJardine) [inline src/models.py:247]: Perhaps make this comment also relate to perplexity? The comments can be better used here.

File: src/models.py [hunk 3]
```diff
@@ -245,6 +259,16 @@ def validation_step(self, batch: Tensor, batch_idx: int):
...
+            name="val_perplexity", 
+            value=perplexity, 
+            prog_bar=True,
+            logger=True, 
+            on_step=False, 
+            on_epoch=True,
+            sync_dist=True,
+            add_dataloader_idx=True)
 
         return loss
```
Comment (KimballNJardine) [inline src/models.py:272]: Do we only want to return loss? I think this is fine as this is for validation (and training up above), but it feels a bit odd to track loss and perplexity and only return loss.

File: src/models.py [hunk 4]
```diff
@@ -106,6 +108,16 @@ def validation_step(self, batch: Tensor, batch_idx: int):
...
```
Comment (KimballNJardine) [inline src/models.py:122]: See comment below about returning loss and perplexity. Probably not needed as this is for training, but it still looks somewhat odd.

File: configs/template_config.yaml [hunk 5]
```diff
@@ -146,7 +148,5 @@ layers: 2
 model_type: "retnet" # Choices: "retnet", "transformer"
 # Sequence Length (int): Context window size by number of tokens
 seq_len: 128
-# Value Embedding Dimension (int): Value embed dimension size
-value_embed_dim: 12
```
Comment (KimballNJardine) [inline configs/template_config.yaml:150]: value_embed_dim is referenced twice in models.py. If this is to be removed, it should be removed there as well. Check that does not cause further problems. I have been looking into if decoder_value_embed_dim does anything, but I am not sure, so testing on a small model to make sure it runs may help.

**Review by KimballNJardine — CHANGES_REQUESTED** [review]
My biggest question involves value_embed_dim. It looks like it has no purpose, but it is referenced elsewhere, notably in models.py. Please look at comments and revise.

File: src/train_model.py (hunk 1 above)
Comment (JayOrten) [reply src/train_model.py:24]: yup

File: src/models.py (hunk 2 above)
Comment (JayOrten) [reply src/models.py:247]: Honestly, I think the code itself is so straightforward that the comments here aren't really necessary

File: src/models.py (hunk 3 above)
Comment (JayOrten) [reply src/models.py:272]: This function is called by torch lightning, and so it must return loss so that lightning can backpropograte, etc. It does not need to return perplexity, we are simply calculating it and then logging it directly.

File: configs/template_config.yaml (hunk 5 above)
Comment (JayOrten) [reply configs/template_config.yaml:150]: It looks like it's used by the retnet, but not transformer. I will add it back in to the config, with a better default value, and remove it for the transformer code.

**Review by KimballNJardine — APPROVED** [review]
Looks good.
Approved.
